### PR TITLE
Interface Evolution

### PR DESCRIPTION
**Pull request by tete17**
This Commits addresses the issues with the current interface. Specifically issues #61 #75 and #78 

My only concern is regarding the meta programming structure `requires_iterators<T>`. Since it is already variadic i propose to substitute the current `is_iterator<T>` and `requires_iterator<T>` with the new ones.

For example in map.h line 54-56, my propose is to change it to a single line =>
`requires_iterators<InputIterators...,InputIt,OutputIt> = 0, `
We can discuss exact wording of the names to be use if you like this idea

### CODE REVIEW

File: include/map.h [hunk 1]
```diff
@@ -36,6 +36,61 @@ namespace grppi {
 @{
 */
 
+/**
+\brief Invoke \ref md_map on a data sequence.
+\tparam InputIterators Iterator types used for input sequences.
+\tparam InputIt Iterator type used for any input sequence.
+\tparam OtuputIt Iterator type used for the output sequence.
+\tparam Transformer Callable type for the transformation operation.
+\param ex Sequential execution policy object.
```
Comment (jdgarciauc3m) [inline include/map.h]: Should be:

ex Execution policy object.

File: include/map.h [hunk 2]
```diff
@@ -36,6 +36,61 @@ namespace grppi {
...
+}
+
+
+
+/**
+\brief Invoke \ref md_map on a data sequence.
+\tparam InputIterators Iterator types used for input sequences.
+\tparam OtuputIt Iterator type used for the output sequence.
+\tparam Transformer Callable type for the transformation operation.
+\param ex Sequential execution policy object.
```
Comment (jdgarciauc3m) [inline include/map.h]: Should be:

ex Execution policy object.

File: include/map.h [hunk 3]
```diff
@@ -36,6 +36,61 @@ namespace grppi {
...
+\tparam InputIterators Iterator types used for input sequences.
+\tparam InputIt Iterator type used for any input sequence.
+\tparam OtuputIt Iterator type used for the output sequence.
+\tparam Transformer Callable type for the transformation operation.
+\param ex Sequential execution policy object.
+\param firsts Tuple of Iterators to the first element in the inputs sequences.
+\param last Iterator to one past the end of one of the input sequence.
+\param first_out Iterator to first element of the output sequence.
+\param transform_op Transformation operation.
+\note The sequential_execution object acts exclusively as a tag type.
```
Comment (jdgarciauc3m) [inline include/map.h]: ????

File: include/map.h [hunk 4]
```diff
@@ -36,6 +36,61 @@ namespace grppi {
...
+\brief Invoke \ref md_map on a data sequence.
+\tparam InputIterators Iterator types used for input sequences.
+\tparam OtuputIt Iterator type used for the output sequence.
+\tparam Transformer Callable type for the transformation operation.
+\param ex Sequential execution policy object.
+\param firsts Tuple of Iterators to the first element in the inputs sequences.
+\param range Size of the input sequences to go through.
+\param first_out Iterator to first element of the output sequence.
+\param transform_op Transformation operation.
+\note The sequential_execution object acts exclusively as a tag type.
```
Comment (jdgarciauc3m) [inline include/map.h]: ????

File: include/map.h [hunk 5]
```diff
@@ -80,7 +135,9 @@ template <typename Execution, typename InputIt, typename OutputIt, typename Tran
           typename ... OtherInputIts,
           requires_iterator<InputIt> = 0,
           requires_iterator<OutputIt> = 0>
-void map(const Execution & ex, 
+[[deprecated("This version of the interface is deprecated.\n"
+             "If you want tu use multiple inputs, use a tuple instead.")]]
```
Comment (jdgarciauc3m) [inline include/map.h]: s/tu/to/

File: include/mapreduce.h [hunk 6]
```diff
@@ -86,7 +151,9 @@ template <typename Execution, typename InputIterator, typename Identity,
           typename Transformer, typename Combiner,
           typename ... OtherInputIterators,
           requires_iterator<InputIterator> = 0>
-auto map_reduce(const Execution & ex, 
+[[deprecated("This version of the interface is deprecated.\n"
+             "If you want tu use multiple inputs, use a tuple instead.")]]
```
Comment (jdgarciauc3m) [inline include/mapreduce.h]: s/tu/to/

File: include/stencil.h [hunk 7]
```diff
@@ -92,6 +160,8 @@ template <typename Execution, typename InputIt, typename OutputIt,
           typename ... OtherInputIts,
           requires_iterator<InputIt> = 0,
           requires_iterator<OutputIt> = 0>
+[[deprecated("This version of the interface is deprecated.\n"
+             "If you want tu use multiple inputs, use a tuple instead.")]]
```
Comment (jdgarciauc3m) [inline include/stencil.h]: s/tu/to/

File: samples/farm/capitalize/main.cpp [hunk 8]
```diff
@@ -49,7 +49,7 @@ void capitalize(grppi::dynamic_execution & e,
     grppi::farm(4,
       [](auto word) { 
         grppi::sequential_execution seq{};
-        grppi::map(seq, begin(word), end(word), begin(word), [](char c) { 
+        grppi::map(seq, std::make_tuple(begin(word)), end(word), begin(word), [](char c) {
```
Comment (jdgarciauc3m) [inline samples/farm/capitalize/main.cpp]: This case should not need a tuple. Isn't it?

**Review by jdgarciauc3m — CHANGES_REQUESTED** [review]
Besides the specific comments, we need to add the corresponding test in the unit tests directory.

File: include/map.h (hunk 3 above)
Comment (tete17) [reply include/map.h]: I suppose you refer to the `\note ` section.

I copy pasted from line 132. Should i remove it from there too?

**Comment by tete17** [comment]
Regarding the unit test, i assume you refer to have a test for every single interface.

Any suggestion on the name of the new member functions responsible for executing the actual test.
For example in map we now have `run_unary` & `run_nary`.

**Comment by jdgarciauc3m** [comment]
Posssible names:

map(ex, make_tuple(i1, i2, i3), e1, o1, transf) -> _nary_tuple_range
map(ex, make_tuple(i1, i2, i3), sz, o1, transf) -> _nary_tuple_size

**Review by Drioastorga — APPROVED** [review]
Ok

**Review by mdolz — APPROVED** [review]
Ok.

**Review by jdgarciauc3m — APPROVED** [review]

**Review by javidoc00 — APPROVED** [review]
Ok

**Pull request merged**